File: sdk/python/rcf_core/sigma.py
# NOTICE: This file is protected under RCF-PL
# [RCF:PROTECTED]
"""
Σ loader — the single point that reads SPECIFICATION/sigma.json and exposes the
semantic alphabet to the rest of rcf_core.

Per RCF-SIGMA.md:
  - sigma.json is the source of truth (this module never hardcodes the alphabet).
  - sigma_version is a "growth ring" of RCF (tracks the protocol version).
  - alphabet_hash is the TRUE comparability key: two fingerprints are comparable
    iff their alphabet_hash matches, even across differing sigma_version strings.

The loader recomputes alphabet_hash from the node/edge structure and refuses to
load if the stored hash disagrees — a tamper / drift tripwire.
"""

from __future__ import annotations

import hashlib
import json
from dataclasses import dataclass
from pathlib import Path


class SigmaError(Exception):
    """Raised when sigma.json is missing, malformed, or its alphabet_hash drifted."""


# Resolve SPECIFICATION/sigma.json relative to this file:
# sdk/python/rcf_core/sigma.py -> repo_root/SPECIFICATION/sigma.json
_REPO_ROOT = Path(__file__).resolve().parents[3]
_DEFAULT_SIGMA_PATH = _REPO_ROOT / "SPECIFICATION" / "sigma.json"


def compute_alphabet_hash(nodes: dict, edges: dict) -> str:
    """
    Canonical alphabet hash. Hashes ONLY the label structure — node classes with
    their sorted ops, and the sorted edge types. Excludes sigma_version and every
    doc string, so a docs-only PATCH release does not change it.

    Must stay byte-identical to the procedure documented in sigma.json/alphabet_hash_doc.
    """
    alphabet = {
        "nodes": {cls: sorted(spec["ops"]) for cls, spec in nodes.items()},
        "edges": sorted(edges.keys()),
    }
    blob = json.dumps(alphabet, sort_keys=True, separators=(",", ":")).encode()
    return "sha256:" + hashlib.sha256(blob).hexdigest()


@dataclass(frozen=True)
class Sigma:
    """An immutable, validated view of the Σ alphabet."""

    version: str
    alphabet_hash: str
    nodes: dict          # class -> {"ops": [...], "doc": str}
    edges: dict          # type  -> {"doc": str, "normative": bool}

    # ─── label validation ────────────────────────────────────────────────────

    def is_class(self, cls: str) -> bool:
        return cls in self.nodes

    def is_label(self, cls: str, op: str | None = None) -> bool:
        """A label is valid if the class exists and (op is None or op ∈ class.ops)."""
        if cls not in self.nodes:
            return False
        if op is None:
            return True
        return op in self.nodes[cls]["ops"]

    def is_edge(self, edge_type: str) -> bool:
        return edge_type in self.edges

    def normative_edges(self) -> list[str]:
        return [t for t, spec in self.edges.items() if spec.get("normative")]

    def require_label(self, cls: str, op: str | None = None) -> None:
        """Raise SigmaError if (cls, op) is not in Σ — used by normalizers."""
        if not self.is_label(cls, op):
            raise SigmaError(f"label not in Σ {self.version}: ({cls!r}, {op!r})")

    # ─── introspection ───────────────────────────────────────────────────────

    def classes(self) -> list[str]:
        return list(self.nodes.keys())

    def ops(self, cls: str) -> list[str]:
        return list(self.nodes[cls]["ops"])


def load_sigma(path: str | Path | None = None, *, verify_hash: bool = True) -> Sigma:
    """
    Load and validate the Σ alphabet from sigma.json.

    Raises SigmaError on missing file, bad JSON, missing required keys, or — when
    verify_hash is True — on alphabet_hash drift between the stored value and the
    value recomputed from the node/edge structure.
    """
    # Resolve to an absolute, normalized path and constrain the input: the Σ
    # alphabet is only ever a .json file. This rejects traversal-style input and
    # non-JSON targets before any read happens.
    p = (Path(path) if path is not None else _DEFAULT_SIGMA_PATH).resolve()
    if p.suffix.lower() != ".json":
        raise SigmaError(f"Σ source must be a .json file, got: {p.name}")
    if not p.is_file():
        raise SigmaError(f"sigma.json not found at: {p}")

    try:
        data = json.loads(p.read_text(encoding="utf-8"))
    except json.JSONDecodeError as e:
        raise SigmaError(f"sigma.json is not valid JSON: {e}") from e

    for key in ("sigma_version", "alphabet_hash", "nodes", "edges"):
        if key not in data:
            raise SigmaError(f"sigma.json missing required key: {key!r}")

    nodes, edges = data["nodes"], data["edges"]
    if not isinstance(nodes, dict) or not isinstance(edges, dict):
        raise SigmaError("sigma.json 'nodes'/'edges' must be objects")

    for cls, spec in nodes.items():
        if not isinstance(spec, dict) or "ops" not in spec or not isinstance(spec["ops"], list):
            raise SigmaError(f"sigma.json node {cls!r} must have an 'ops' list")

    recomputed = compute_alphabet_hash(nodes, edges)
    if verify_hash and recomputed != data["alphabet_hash"]:
        raise SigmaError(
            "alphabet_hash drift — sigma.json was edited without recomputing the hash.\n"
            f"  stored    : {data['alphabet_hash']}\n"
            f"  recomputed: {recomputed}\n"
            "Either revert the change or update alphabet_hash (this is a breaking Σ change)."
        )

    return Sigma(
        version=data["sigma_version"],
        alphabet_hash=data["alphabet_hash"],
        nodes=nodes,
        edges=edges,
    )

File: sdk/typescript/src/core/sigma.ts
// NOTICE: This file is protected under RCF-PL
// [RCF:PROTECTED]

import { fileURLToPath } from 'url';
import { dirname, join, resolve } from 'path';
import { readFileSync, existsSync } from 'fs';
import { createHash, timingSafeEqual } from 'crypto';

/**
 * Constant-time string comparison for hash values.
 * Prevents timing side-channel attacks on alphabet hash checks.
 */
export function timingSafeHashEqual(a: string, b: string): boolean {
  const bufA = Buffer.from(a, 'utf-8');
  const bufB = Buffer.from(b, 'utf-8');
  if (bufA.length !== bufB.length) return false;
  return timingSafeEqual(bufA, bufB);
}

export class SigmaError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'SigmaError';
  }
}

// Find repository root from current file context
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const REPO_ROOT = resolve(__dirname, '../../../../');
const DEFAULT_SIGMA_PATH = join(REPO_ROOT, 'SPECIFICATION', 'sigma.json');

export function computeAlphabetHash(nodes: Record<string, any>, edges: Record<string, any>): string {
  const alphabet: Record<string, any> = {
    nodes: {},
    edges: Object.keys(edges).sort()
  };

  for (const [cls, spec] of Object.entries(nodes)) {
    alphabet.nodes[cls] = [...spec.ops].sort();
  }

  // Deterministic JSON stringify without whitespace
  const sortedKeys = Object.keys(alphabet).sort();
  const canonicalForm: Record<string, any> = {};
  for (const k of sortedKeys) {
    if (k === 'nodes') {
      const sortedNodes: Record<string, string[]> = {};
      for (const cls of Object.keys(alphabet.nodes).sort()) {
        sortedNodes[cls] = alphabet.nodes[cls];
      }
      canonicalForm[k] = sortedNodes;
    } else {
      canonicalForm[k] = alphabet[k];
    }
  }

  const blob = JSON.stringify(canonicalForm);
  const hash = createHash('sha256').update(blob).digest('hex');
  return 'sha256:' + hash;
}

export class Sigma {
  constructor(
    public readonly version: string,
    public readonly alphabetHash: string,
    public readonly nodes: Record<string, { ops: string[]; doc: string }>,
    public readonly edges: Record<string, { doc: string; normative: boolean }>
  ) {}

  public isClass(cls: string): boolean {
    return cls in this.nodes;
  }

  public isLabel(cls: string, op?: string): boolean {
    if (!(cls in this.nodes)) {
      return false;
    }
    if (op === undefined) {
      return true;
    }
    return this.nodes[cls].ops.includes(op);
  }

  public isEdge(edgeType: string): boolean {
    return edgeType in this.edges;
  }

  public normativeEdges(): string[] {
    return Object.entries(this.edges)
      .filter(([_, spec]) => spec.normative)
      .map(([type]) => type);
  }

  public requireLabel(cls: string, op?: string): void {
    if (!this.isLabel(cls, op)) {
      throw new SigmaError(`label not in Σ ${this.version}: (${cls}, ${op})`);
    }
  }

  public classes(): string[] {
    return Object.keys(this.nodes);
  }

  public ops(cls: string): string[] {
    return [...this.nodes[cls].ops];
  }
}

export function loadSigma(filePath?: string, verifyHash = true): Sigma {
  const p = filePath ? resolve(filePath) : DEFAULT_SIGMA_PATH;
  if (!p.endsWith('.json')) {
    throw new SigmaError(`Σ source must be a .json file, got: ${p}`);
  }
  if (!existsSync(p)) {
    throw new SigmaError(`sigma.json not found at: ${p}`);
  }

  let data: any;
  try {
    data = JSON.parse(readFileSync(p, 'utf-8'));
  } catch (e: any) {
    throw new SigmaError(`sigma.json is not valid JSON: ${e.message}`);
  }

  for (const key of ['sigma_version', 'alphabet_hash', 'nodes', 'edges']) {
    if (!(key in data)) {
      throw new SigmaError(`sigma.json missing required key: '${key}'`);
    }
  }

  const { nodes, edges } = data;
  if (typeof nodes !== 'object' || typeof edges !== 'object' || nodes === null || edges === null) {
    throw new SigmaError("sigma.json 'nodes'/'edges' must be objects");
  }

  for (const [cls, spec] of Object.entries(nodes)) {
    if (typeof spec !== 'object' || spec === null || !('ops' in spec) || !Array.isArray((spec as any).ops)) {
      throw new SigmaError(`sigma.json node '${cls}' must have an 'ops' list`);
    }
  }

  const recomputed = computeAlphabetHash(nodes, edges);
  if (verifyHash && !timingSafeHashEqual(recomputed, data.alphabet_hash)) {
    throw new SigmaError(
      `alphabet_hash drift — sigma.json was edited without recomputing the hash.\n` +
      `  stored    : [redacted]\n` +
      `  recomputed: [redacted]\n` +
      `Either revert the change or update alphabet_hash (this is a breaking Σ change).`
    );
  }

  return new Sigma(
    data.sigma_version,
    data.alphabet_hash,
    nodes,
    edges
  );
}
